src/decoder/mp3.rs

```diff
@@ -1,129 +1,917 @@
+//!
+//! - Duration scanning can be slow for large files
+//! - Variable bitrate files may have less accurate seeking
+//! - Frame-based decoding minimizes memory usage
+//!
+//! # Example
+//!
+//! ```ignore
+//! use std::fs::File;
+//! use rodio::Decoder;
+//!
+//! let file = File::open("audio.mp3").unwrap();
+//! let decoder = Decoder::builder()
+//!     .with_data(file)
+//!     .with_seekable(true)
+//!     .with_scan_duration(true)
+//!     .build()
+//!     .unwrap();
+//!
+//! // MP3 format doesn't support bit depth detection
+//! assert_eq!(decoder.bits_per_sample(), None);
+//! ```
 
-use crate::common::{ChannelCount, Sample, SampleRate};
-use crate::source::SeekError;
-use crate::Source;
+use std::{
+    io::{Read, Seek, SeekFrom},
+    num::NonZero,
+    sync::Arc,
+    time::Duration,
+};
 
 use dasp_sample::Sample as _;
 
-use minimp3::Decoder;
-use minimp3::Frame;
+use minimp3::{Decoder, Frame};
 use minimp3_fixed as minimp3;
 
+use super::{utils, Settings};
+use crate::{
+    common::{ChannelCount, Sample, SampleRate},
+    decoder::builder::SeekMode,
+    source::SeekError,
+    Source,
+};
+
+/// Decoder for the MP3 format using the `minimp3` library.
+///
+/// Provides lossy audio decoding with frame-based processing and coarse seeking support.
+/// Duration calculation may require file scanning for accurate results with variable bitrate files.
+///
+/// # Frame-based Processing
+///
+/// MP3 audio is organized into variable-size frames containing compressed audio data.
+/// Each frame is decoded independently, containing 384 (Layer I) or 1152 (Layer II/III)
+/// samples per channel. The decoder maintains the current frame and tracks position
+/// within it for efficient sample-by-sample iteration.
+///
+/// # Seeking Behavior
+///
+/// MP3 seeking accuracy depends on the encoding type:
+/// - **Constant Bitrate (CBR)**: Accurate byte-position-based seeking
+/// - **Variable Bitrate (VBR)**: Approximate seeking with potential drift
+/// - **Average Bitrate (ABR)**: Moderate accuracy depending on variation
+///
+/// # Bitrate Adaptation
+///
+/// The decoder tracks average bitrate over time to improve seeking accuracy,
+/// especially for VBR files where initial estimates may be inaccurate.
+///
+/// # Channel Changes
+///
+/// MP3 frames can theoretically change channel configuration, though this is
+/// rare in practice. The decoder handles such changes dynamically.
+///
+/// # Thread Safety
+///
+/// This decoder is not thread-safe. Create separate instances for concurrent access
+/// or use appropriate synchronization primitives.
+///
+/// # Generic Parameters
+///
+/// * `R` - The underlying data source type, must implement `Read + Seek`
 pub struct Mp3Decoder<R>
 where
     R: Read + Seek,
 {
-    // decoder: SeekDecoder<R>,
-    decoder: Decoder<R>,
-    // what minimp3 calls frames rodio calls spans
-    current_span: Frame,
+    /// The underlying minimp3 decoder, wrapped in Option for seeking operations.
```

Comment:
The fields are private, so they (and these docs) will not show on docs.rs. I assume we want these fields to stay private so we can write them way less verbose (increase the signal to noise ratio).

This could be: `Option to allow owning the Decoder instance during seeking.`